lib/backend/k8s/resources/customresource.go

```diff
@@ -263,62 +267,77 @@ func (c *customK8sResourceClient) List(ctx context.Context, list model.ListInter
 		}
 	}
 
-	// Since we are not performing an exact Get, Kubernetes will return a
-	// list of resources.
-	reslOut := reflect.New(c.k8sListType).Interface().(ResourceList)
-
 	// If it is a namespaced resource, then we'll need the namespace.
 	namespace := list.(model.ResourceListOptions).Namespace
 
-	// Perform the request.
-	err := c.restClient.Get().
-		NamespaceIfScoped(namespace, c.namespaced).
-		Resource(c.resource).
-		Param("resourceVersion", revision).
-		Do(ctx).Into(reslOut)
-	if err != nil {
-		// Don't return errors for "not found".  This just
-		// means there are no matching Custom K8s Resources, and we should return
-		// an empty list.
-		if !kerrors.IsNotFound(err) {
-			log.WithError(err).Debug("Error listing resources")
-			return nil, K8sErrorToCalico(err, list)
+	// listFunc performs a list with the given options.
+	listFunc := func(ctx context.Context, opts metav1.ListOptions) (runtime.Object, error) {
+		out := reflect.New(c.k8sListType).Interface().(ResourceList)
+		err := c.restClient.Get().
+			NamespaceIfScoped(namespace, c.namespaced).
+			Resource(c.resource).
+			VersionedParams(&opts, scheme.ParameterCodec).
+			Do(ctx).Into(out)
+		if err != nil {
+			// Don't return errors for "not found".  This just
+			// means there are no matching Custom K8s Resources, and we should return
+			// an empty list.
+			if !kerrors.IsNotFound(err) {
+				logContext.WithError(err).Debug("Error listing resources")
+				return nil, err
+			}
 		}
-		return &model.KVPairList{
-			KVPairs:  kvps,
-			Revision: revision,
-		}, nil
+		return out, nil
 	}
 
-	// We expect the list type to have an "Items" field that we can
-	// iterate over.
-	elem := reflect.ValueOf(reslOut).Elem()
-	items := reflect.ValueOf(elem.FieldByName("Items").Interface())
-	for idx := 0; idx < items.Len(); idx++ {
-		res := items.Index(idx).Addr().Interface().(Resource)
-		if kvp, err := c.convertResourceToKVPair(res); err == nil {
-			kvps = append(kvps, kvp)
-		} else {
+	lp := pager.New(listFunc)
+	opts := metav1.ListOptions{ResourceVersion: revision}
+	if revision != "" {
```

Comment:
I think this is the right way to do this. There are three scenarios:

- revision = "" -> resourceVersionMatch not supported.
- revision = "0" -> resourceVersionMatch=NotOlderThan -> Any (from API server cache)
- revision = <non-zero> -> resourceVersionMatch=NotOlderThan -> Not older than the given value.